geode-redis/src/distributedTest/java/org/apache/geode/redis/session/SessionsAndCrashesDUnitTest.java

```diff
@@ -0,0 +1,236 @@
+    locatorProperties.setProperty(MAX_WAIT_TIME_RECONNECT, "15000");
+
+    locator = cluster.startLocatorVM(0, locatorProperties);
+
+    server1 = startRedisVM(1, 0);
+    server2 = startRedisVM(2, 0);
+    server3 = startRedisVM(3, 0);
+
+    redisPorts = new int[] {
+        cluster.getRedisPort(1),
+        cluster.getRedisPort(2),
+        cluster.getRedisPort(3)};
+
+    jedis = new Jedis("localhost", redisPorts[0], JEDIS_TIMEOUT);
+  }
+
+  private static MemberVM startRedisVM(int vmId, Integer redisPort) {
+    int locatorPort = locator.getPort();
+
+    return cluster.startRedisVM(vmId, x -> x
+        .withProperty(REDIS_PORT, redisPort.toString())
+        .withConnectionToLocator(locatorPort));
+  }
+
+  @Before
+  @SuppressWarnings("unchecked")
+  public void setup() {
+    String[] args = new String[] {
+        "" + redisPorts[2],
+        "" + redisPorts[1],
+        "" + redisPorts[0]};
+
+    springContext = SpringApplication.run(RedisSpringTestApplication.class, args);
+    sessionRepository = springContext.getBean(SessionRepository.class);
+    assertThat(sessionRepository).isNotNull();
+  }
+
+  @After
+  public void teardown() {
+    springContext.stop();
+    jedis.flushAll();
+    sessionIds.clear();
+  }
+
+  @Test
+  public void sessionOperationsDoNotFail_whileServersAreRestarted() throws Exception {
+    createSessions();
+
+    AtomicBoolean running = new AtomicBoolean(true);
+    AtomicReference<String> phase = new AtomicReference<>("STARTUP");
+
+    Future<Integer> future1 = executor.submit(() -> sessionUpdater(0, running, phase));
+    Future<Integer> future2 = executor.submit(() -> sessionUpdater(1, running, phase));
+
+    GeodeAwaitility.await().during(1, TimeUnit.SECONDS).until(() -> true);
+
+    phase.set("CRASH 1 SERVER2");
+    cluster.crashVM(2);
+    server2 = startRedisVM(2, redisPorts[1]);
+    phase.set("CRASH 1 REBALANCING");
+    rebalanceAllRegions(server2);
+
+    phase.set("CRASH 2 SERVER3");
+    cluster.crashVM(3);
+    server3 = startRedisVM(3, redisPorts[2]);
+    phase.set("CRASH 2 REBALANCING");
+    rebalanceAllRegions(server3);
+
+    phase.set("CRASH 3 SERVER2");
+    cluster.crashVM(2);
+    server2 = startRedisVM(2, redisPorts[1]);
+    phase.set("CRASH 3 REBALANCING");
+    rebalanceAllRegions(server2);
+
+    phase.set("CRASH 4 SERVER3");
+    cluster.crashVM(3);
+    server3 = startRedisVM(3, redisPorts[2]);
+    phase.set("CRASH 4 REBALANCING");
+    rebalanceAllRegions(server3);
+
+    phase.set("FINISHING");
+    GeodeAwaitility.await().during(10, TimeUnit.SECONDS).until(() -> true);
+
+    running.set(false);
+    future1.get();
+    future2.get();
+  }
+
+  private Integer sessionUpdater(int index, AtomicBoolean running, AtomicReference<String> phase) {
+    int count = 0;
+    while (running.get()) {
+      int modCount = count % NUM_SESSIONS;
+      String sessionId = sessionIds.get(modCount);
+      Session session = findSession(sessionId);
```

Comment:
between this and the call to saveSession(), the sessionUpdater method could take over 40 seconds. the wait in the test for the finish phase only waits 10 seconds, so a call to sessionUpdater could potentially be cut short.